File: common/src/main/java/com/obsidian/backup/common/IpcClient.java
package com.obsidian.backup.common;

import java.io.*;
import java.net.UnixDomainSocketAddress;
import java.nio.channels.*;
import java.nio.file.Path;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Consumer;

/**
 * Loader-agnostic IPC client for communicating with the Obsidian Sidecar daemon
 * via Unix Domain Socket (UDS), using Java 16+ native support.
 *
 * Thread safety: uses AtomicBoolean for the connected flag to ensure
 * visibility across the read thread and the main server thread.
 */
public class IpcClient implements AutoCloseable {

    private final String socketPath;
    private final String authToken;
    private SocketChannel channel;
    private BufferedWriter writer;
    private BufferedReader reader;
    private final AtomicBoolean connected = new AtomicBoolean(false);

    private final ConcurrentHashMap<String, Consumer<IpcProtocol.Response>> pendingRequests = new ConcurrentHashMap<>();
    private final ConcurrentLinkedQueue<String> responseQueue = new ConcurrentLinkedQueue<>();

    private volatile Thread readThread;
    private final Logger logger;

    public interface Logger {
        void info(String msg, Object... args);
        void warn(String msg, Object... args);
        void error(String msg, Object... args);
    }

    public IpcClient(String socketPath, String authToken, Logger logger) {
        this.socketPath = socketPath;
        this.authToken = authToken;
        this.logger = logger;
    }

    /** Connect to the Sidecar daemon and authenticate. */
    public boolean connect() {
        if (isConnected()) return true;

        try {
            var address = UnixDomainSocketAddress.of(Path.of(socketPath));
            channel = SocketChannel.open(address);
            channel.configureBlocking(true);

            writer = new BufferedWriter(new OutputStreamWriter(Channels.newOutputStream(channel)));
            reader = new BufferedReader(new InputStreamReader(Channels.newInputStream(channel)));

            // Authenticate
            var authRequest = new IpcProtocol.Request(
                IpcProtocol.OpCode.AUTH.code,
                IpcProtocol.paramsAuth(authToken != null ? authToken : "obsidian-default-token")
            );
            String authJson = IpcProtocol.toJson(authRequest);
            writer.write(authJson);
            writer.newLine();
            writer.flush();

            String authLine = reader.readLine();
            if (authLine == null) {
                logger.error("[IPC] Auth failed: no response from sidecar");
                close();
                return false;
            }

            var authResp = IpcProtocol.parseResponse(authLine);
            if (!"ok".equals(authResp.status)) {
                logger.error("[IPC] Auth rejected: {}", authResp.message);
                close();
                return false;
            }

            connected.set(true);
            logger.info("[IPC] Connected and authenticated at {}", socketPath);
            startReadLoop();
            return true;

        } catch (Exception e) {
            logger.error("[IPC] Connection failed: {}", e.getMessage());
            return false;
        }
    }

    /** Gracefully disconnect. */
    public void close() {
        connected.set(false);
        var thread = readThread;
        readThread = null;
        if (thread != null) {
            thread.interrupt();
            try { thread.join(500); }
            catch (InterruptedException e) { Thread.currentThread().interrupt(); }
        }
        try { writer.close(); } catch (Exception ignored) {}
        try { reader.close(); } catch (Exception ignored) {}
        try { channel.close(); } catch (Exception ignored) {}
        writer = null; reader = null; channel = null;
        logger.info("[IPC] Disconnected");
    }

    private void startReadLoop() {
        readThread = new Thread(() -> {
            try {
                while (connected.get()) {
                    try {
                        String line = reader.readLine();
                        if (line == null) {
                            logger.warn("[IPC] Sidecar closed connection (EOF)");
                            break;
                        }
                        if (line.isBlank()) continue;
                        responseQueue.add(line);
                    } catch (InterruptedException | ClosedChannelException e) {
                        break;
                    } catch (Exception e) {
                        if (connected.get()) logger.error("[IPC] Read error: {}", e.getMessage());
                        break;
                    }
                }
            } catch (Exception e) {
                if (connected.get()) logger.error("[IPC] Read loop crashed: {}", e.getMessage());
            } finally {
                connected.set(false);
            }
        }, "Obsidian-IPC-Reader");
        readThread.setDaemon(true);
        readThread.start();
    }

    /** Called each tick to dispatch queued responses on the main thread. */
    public void pollResponses() {
        String raw;
        while ((raw = responseQueue.poll()) != null) {
            try {
                var response = IpcProtocol.parseResponse(raw);
                var callback = pendingRequests.remove(response.tx_id);
                if (callback != null) callback.accept(response);
            } catch (Exception e) {
                logger.error("[IPC] Failed to parse response: {}", e.getMessage());
            }
        }
    }

    /** Send an async request with a callback. */
    public boolean sendRequest(IpcProtocol.OpCode op, Map<String, Object> params,
                                Consumer<IpcProtocol.Response> callback) {
        if (!isConnected()) {
            logger.warn("[IPC] Not connected");
            return false;
        }

        var request = new IpcProtocol.Request(op.code, params);
        pendingRequests.put(request.tx_id, callback);

        try {
            String json = IpcProtocol.toJson(request);
            synchronized (writer) {
                writer.write(json);
                writer.newLine();
                writer.flush();
            }
            return true;
        } catch (Exception e) {
            pendingRequests.remove(request.tx_id);
            logger.error("[IPC] Send failed: {}", e.getMessage());
            return false;
        }
    }

    public boolean isConnected() {
        return connected.get();
    }
}

File: mod-neoforge/src/main/kotlin/com/obsidian/backup/ipc/IpcClient.kt
package com.obsidian.backup.ipc

import com.obsidian.backup.config.ModConfig
import com.obsidian.backup.ObsidianBackupMod
import net.minecraft.network.chat.Component
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.net.UnixDomainSocketAddress
import java.nio.channels.Channels
import java.nio.channels.ClosedChannelException
import java.nio.channels.SocketChannel
import java.nio.file.Path
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.atomic.AtomicBoolean
import java.util.function.Consumer

/**
 * IPC client that communicates with the Obsidian Sidecar daemon
 * via Unix Domain Socket (UDS) using Java 16+ native support.
 *
 * Thread safety: uses AtomicBoolean for the connected flag to ensure
 * visibility across the read thread and the main server thread.
 * The read loop handles all known channel-closure exceptions explicitly.
 */
class IpcClient(private val config: ModConfig) {

    private var channel: SocketChannel? = null
    private var writer: BufferedWriter? = null
    private var reader: BufferedReader? = null

    // AtomicBoolean ensures the read thread sees disconnection immediately
    private val connected = AtomicBoolean(false)

    private val pendingRequests = ConcurrentHashMap<String, Consumer<IpcProtocol.Response>>()
    private val responseQueue = ConcurrentLinkedQueue<String>()

    @Volatile
    private var readThread: Thread? = null

    /**
     * Connect to the Sidecar daemon via UDS.
     */
    fun connect(): Boolean {
        if (isConnected()) return true

        return try {
            val socketPath = Path.of(config.sidecarSocketPath)
            val address = UnixDomainSocketAddress.of(socketPath)
            channel = SocketChannel.open(address).apply {
                configureBlocking(true)
            }

            writer = BufferedWriter(OutputStreamWriter(Channels.newOutputStream(channel)))
            reader = BufferedReader(InputStreamReader(Channels.newInputStream(channel)))

            connected.set(true)
            ObsidianBackupMod.LOGGER.info("[IPC] Connected to Sidecar at {}", config.sidecarSocketPath)

            startReadLoop()
            true
        } catch (e: Exception) {
            ObsidianBackupMod.LOGGER.error("[IPC] Failed to connect to Sidecar: {}", e.message)
            false
        }
    }

    /**
     * Disconnect from the Sidecar gracefully.
     *
     * Order of operations is important:
     * 1. Set connected = false (stops the read loop)
     * 2. Interrupt the read thread (breaks any blocking read)
     * 3. Wait briefly for the thread to exit
     * 4. Close resources
     */
    fun disconnect() {
        connected.set(false)
        val thread = readThread
        readThread = null
        thread?.interrupt()

        // Give the read thread a moment to exit gracefully
        try {
            thread?.join(500)
        } catch (_: InterruptedException) {
            Thread.currentThread().interrupt()
        }

        try {
            writer?.close()
            reader?.close()
            channel?.close()
        } catch (_: Exception) {
            // Best-effort cleanup
        }
        writer = null
        reader = null
        channel = null
        ObsidianBackupMod.LOGGER.info("[IPC] Disconnected from Sidecar")
    }

    /**
     * Background thread that reads responses from the Sidecar.
     *
     * Uses AtomicBoolean.get() for the loop condition to guarantee
     * visibility of disconnection. Handles all known channel-closure
     * exceptions explicitly to avoid silent thread exit.
     */
    private fun startReadLoop() {
        readThread = Thread {
            try {
                while (connected.get()) {
                    try {
                        val line = reader?.readLine()
                        if (line == null) {
                            // EOF — sidecar closed the connection
                            ObsidianBackupMod.LOGGER.warn("[IPC] Sidecar closed connection (EOF)")
                            break
                        }
                        if (line.isBlank()) continue
                        responseQueue.add(line)
                    } catch (e: InterruptedException) {
                        // Expected during disconnect — exit cleanly
                        break
                    } catch (e: ClosedChannelException) {
                        // Channel closed by disconnect() — exit cleanly
                        break
                    } catch (e: Exception) {
                        if (connected.get()) {
                            ObsidianBackupMod.LOGGER.error("[IPC] Read error: {}", e.message)
                        }
                        break
                    }
                }
            } catch (e: Exception) {
                if (connected.get()) {
                    ObsidianBackupMod.LOGGER.error("[IPC] Read loop crashed: {}", e.message)
                }
            } finally {
                // Ensure connected is false if we exited unexpectedly
                connected.set(false)
            }
        }.apply {
            name = "Obsidian-IPC-Reader"
            isDaemon = true
            start()
        }
    }

    /**
     * Called every server tick to process queued responses on the main thread.
     */
    fun pollResponses() {
        while (true) {
            val raw = responseQueue.poll() ?: break
            try {
                val response = IpcProtocol.parseResponse(raw)
                val callback = pendingRequests.remove(response.tx_id)
                callback?.accept(response)
            } catch (e: Exception) {
                ObsidianBackupMod.LOGGER.error("[IPC] Failed to parse response: {}", e.message)
            }
        }
    }

    /**
     * Send a request to the Sidecar and register a callback for the response.
     */
    fun sendRequest(
        op: IpcProtocol.OpCode,
        params: Map<String, Any?>,
        callback: Consumer<IpcProtocol.Response>
    ): Boolean {
        if (!isConnected()) {
            ObsidianBackupMod.LOGGER.warn("[IPC] Not connected to Sidecar")
            return false
        }

        val request = IpcProtocol.Request(op = op.code, params = params)
        pendingRequests[request.tx_id] = callback

        return try {
            val json = IpcProtocol.toJson(request)
            synchronized(writer!!) {
                writer!!.write(json)
                writer!!.newLine()
                writer!!.flush()
            }
            true
        } catch (e: Exception) {
            pendingRequests.remove(request.tx_id)
            ObsidianBackupMod.LOGGER.error("[IPC] Failed to send request: {}", e.message)
            false
        }
    }

    /**
     * Send a request synchronously (blocking, for use off the main thread).
     *
     * NOTE: This bypasses the callback/pollResponse mechanism. The caller
     * must ensure this is not interleaved with async sendRequest calls.
     */
    fun sendRequestSync(op: IpcProtocol.OpCode, params: Map<String, Any?>): IpcProtocol.Response? {
        if (!isConnected()) return null

        val request = IpcProtocol.Request(op = op.code, params = params)

        return try {
            val json = IpcProtocol.toJson(request)
            synchronized(writer!!) {
                writer!!.write(json)
                writer!!.newLine()
                writer!!.flush()
            }
            val responseLine = reader?.readLine() ?: return null
            IpcProtocol.parseResponse(responseLine)
        } catch (e: Exception) {
            ObsidianBackupMod.LOGGER.error("[IPC] Sync request failed: {}", e.message)
            null
        }
    }

    fun isConnected(): Boolean = connected.get()
}
